tools/nmt_grpc_service/server.py

```diff
@@ -0,0 +1,175 @@
+    )
+    parser.add_argument("--port", default=50052, type=int, required=False)
+    parser.add_argument("--batch_size", type=int, default=256, help="Maximum number of batches to process")
+    parser.add_argument("--beam_size", type=int, default=1, help="Beam Size")
+    parser.add_argument("--len_pen", type=float, default=0.6, help="Length Penalty")
+    parser.add_argument("--max_delta_length", type=int, default=5, help="Max Delta Generation Length.")
+
+    args = parser.parse_args()
+    return args
+
+
+def batches(lst, n):
+    """Yield successive n-sized chunks from lst."""
+    for i in range(0, len(lst), n):
+        yield lst[i : i + n]
+
+
+class RivaTranslateServicer(nmtsrv.RivaTranslateServicer):
+    """Provides methods that implement functionality of route guide server."""
+
+    def __init__(
+        self, model_paths, punctuation_model_path, beam_size=1, len_pen=0.6, max_delta_length=5, batch_size=256
+    ):
+        self._models = {}
+        self._beam_size = beam_size
+        self._len_pen = len_pen
+        self._max_delta_length = max_delta_length
+        self._batch_size = batch_size
+        self._punctuation_model_path = punctuation_model_path
+        self._model_paths = model_paths
+
+        for model_path in self._model_paths:
+            assert os.path.exists(model_path)
+            logging.info(f"Loading model {model_path}")
+            self._load_model(model_path)
+
+        if self._punctuation_model_path != "":
+            assert os.path.exists(punctuation_model_path)
+            logging.info(f"Loading punctuation model {model_path}")
+            self._load_puncutation_model(punctuation_model_path)
+
+        logging.info("Models loaded. Ready for inference requests.")
+
+    def _load_puncutation_model(self, punctuation_model_path):
+        if punctuation_model_path.endswith(".nemo"):
+            self.punctuation_model = nemo_nlp.models.PunctuationCapitalizationModel.restore_from(
+                restore_path=punctuation_model_path
+            )
+        else:
+            raise NotImplemented(f"Only support .nemo files, but got: {punctuation_model_path}")
+
+        if torch.cuda.is_available():
+            self.punctuation_model = self.punctuation_model.cuda()
+
+    def _load_model(self, model_path):
+        if model_path.endswith(".nemo"):
+            logging.info("Attempting to initialize from .nemo file")
+            model = nemo_nlp.models.machine_translation.MTEncDecModel.restore_from(restore_path=model_path)
+            model.beam_search.beam_size = self._beam_size
+            model.beam_search.len_pen = self._len_pen
+            model.beam_search.max_delta_length = self._max_delta_length
+        else:
+            raise NotImplemented(f"Only support .nemo files, but got: {model_path}")
+
+        if not hasattr(model, "src_language") or not hasattr(model, "tgt_language"):
+            raise ValueError(f"Could not find src_language and tgt_language in model attributes")
+
```

Comment:
Can we add these to the argparser? Then if a model doesn't have them, it could be updated by the user specified languages?